### PR TITLE
lambda peek optimizer

### PR DESCRIPTION
**Pull request by havardpe**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@baldersheim please review
@lesters FYI

### CODE REVIEW

File: eval/src/vespa/eval/eval/tensor_function.h [hunk 1]
```diff
@@ -343,6 +343,9 @@ class Lambda : public Node
 public:
     Lambda(const ValueType &result_type_in, const std::vector<size_t> &bindings_in, const Function &lambda_in, NodeTypes lambda_types_in)
         : Node(result_type_in), _bindings(bindings_in), _lambda(lambda_in.shared_from_this()), _lambda_types(std::move(lambda_types_in)) {}
+    const std::vector<size_t> bindings() const { return _bindings; }
```
Comment (baldersheim) [inline eval/src/vespa/eval/eval/tensor_function.h:346]: I think you forgot the & to copying the vector.

File: eval/src/vespa/eval/tensor/dense/dense_cell_range_function.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+DenseCellRangeFunction::DenseCellRangeFunction(const eval::ValueType &result_type,
+                                               const eval::TensorFunction &child,
+                                               size_t offset, size_t length)
+    : eval::tensor_function::Op1(result_type, child),
+      _offset(offset),
+      _length(length)
+{
+}
+
+DenseCellRangeFunction::~DenseCellRangeFunction()
```
Comment (baldersheim) [inline eval/src/vespa/eval/tensor/dense/dense_cell_range_function.cpp:42]: Use = default.

**Review by baldersheim — APPROVED** [review]

File: eval/src/vespa/eval/eval/node_tools.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+struct CopyNode : NodeTraverser, NodeVisitor {
+
+    std::unique_ptr<Error> error;
+    std::vector<Node_UP> stack;
+
+    CopyNode() : error(), stack() {}
+    ~CopyNode();
+
+    Node_UP get_result() {
```
Comment (baldersheim) [inline eval/src/vespa/eval/eval/node_tools.cpp:31]: get_ signals non destructive access.

File: eval/src/vespa/eval/tensor/dense/dense_lambda_peek_optimizer.cpp [hunk 4]
```diff
@@ -1,13 +1,204 @@
...
+            if (i == 0) {
+                idx_expr = std::move(dim_expr);
+            } else {
+                idx_expr = make_floor(std::move(idx_expr));
+                idx_expr = make_op<Mul>(std::move(idx_expr), std::make_unique<Number>(src_dim_sizes[i]));
+                idx_expr = make_op<Add>(std::move(idx_expr), std::move(dim_expr));
+            }
+        }
+        src_idx_fun = Function::create(std::move(idx_expr), dst_type.dimension_names());
+        auto str = src_idx_fun->dump_as_lambda();
```
Comment (havardpe) [inline eval/src/vespa/eval/tensor/dense/dense_lambda_peek_optimizer.cpp:114]: remove this

File: eval/src/vespa/eval/eval/tensor_function.h (hunk 1 above)
Comment (havardpe) [reply eval/src/vespa/eval/eval/tensor_function.h:346]: fixed

File: eval/src/vespa/eval/tensor/dense/dense_cell_range_function.cpp (hunk 2 above)
Comment (havardpe) [reply eval/src/vespa/eval/tensor/dense/dense_cell_range_function.cpp:42]: fixed

File: eval/src/vespa/eval/tensor/dense/dense_lambda_peek_function.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+DenseLambdaPeekFunction::DenseLambdaPeekFunction(const ValueType &result_type,
+                                                 const TensorFunction &child,
+                                                 std::shared_ptr<Function const> idx_fun)
+    : Op1(result_type, child),
+      _idx_fun(std::move(idx_fun))
+{
+}
+
+DenseLambdaPeekFunction::~DenseLambdaPeekFunction()
```
Comment (baldersheim) [inline eval/src/vespa/eval/tensor/dense/dense_lambda_peek_function.cpp:72]: = default

File: eval/src/vespa/eval/tensor/dense/dense_cell_range_function.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+{
+private:
+    size_t _offset;
+    size_t _length;
+
+public:
+    DenseCellRangeFunction(const eval::ValueType &result_type,
+                           const eval::TensorFunction &child,
+                           size_t offset, size_t length);
+    ~DenseCellRangeFunction();
```
Comment (baldersheim) [inline eval/src/vespa/eval/tensor/dense/dense_cell_range_function.h:24]: I normally add override on destructors too. clion(clang) likes that.

File: eval/src/vespa/eval/eval/node_tools.cpp (hunk 3 above)
Comment (havardpe) [reply eval/src/vespa/eval/eval/node_tools.cpp:31]: fixed

File: eval/src/vespa/eval/tensor/dense/dense_lambda_peek_optimizer.cpp (hunk 4 above)
Comment (havardpe) [reply eval/src/vespa/eval/tensor/dense/dense_lambda_peek_optimizer.cpp:114]: fixed

**Review by baldersheim — APPROVED** [review]
I think you are good to go for a factory tun.
Apart from the copy vector by value you should be fine.

File: eval/src/vespa/eval/tensor/dense/dense_lambda_peek_function.cpp (hunk 5 above)
Comment (havardpe) [reply eval/src/vespa/eval/tensor/dense/dense_lambda_peek_function.cpp:72]: fixed

File: eval/src/vespa/eval/tensor/dense/dense_cell_range_function.h (hunk 6 above)
Comment (havardpe) [reply eval/src/vespa/eval/tensor/dense/dense_cell_range_function.h:24]: I will try to remember that. Added override to new classes.

**Pull request merged**